Solution to jv-array-list homework

anurmetov opened the pull request.

anurmetov committed:
- Created my own Array List which behaves 1:1 like java.util.ArrayList with all functions inside

anurmetov committed:
- fixed some clean code issues

anurmetov committed:
- fixed some clean code issues

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,164 @@
 package core.basesyntax;
 
+import java.util.Arrays;
```
Line 3: Import of java.util.Arrays. This violates the checklist: "Don't use Objects, Arrays, or any other util class." Remove this import and replace usages (`Arrays.copyOf`, `Arrays.toString`) with your own logic and System.arraycopy-based resizing.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,164 @@
+            int newCapacity = (elementData.length == 0)
+                    ? DEFAULT_CAPACITY : (int) (elementData.length * 1.5);
+            elementData = Arrays.copyOf(elementData, newCapacity);
+        }
```
Lines 27-31 and 39-43: Growth/resizing logic is duplicated and uses Arrays.copyOf. Checklist requires: "Resize the array in a separate method." and "Don't create repeating code." Extract a single private resize/grow method that computes new capacity using a named constant and copies elements using System.arraycopy instead of Arrays.copyOf.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,164 @@
 
+        if (size == elementData.length) {
+            int newCapacity = (elementData.length == 0)
+                    ? DEFAULT_CAPACITY : (int) (elementData.length * 1.5);
```
Lines 29 and 41: The growth factor `1.5` is used as an inline literal. This violates: "Any magic numbers should be constants". Introduce a named constant (e.g. GROWTH_FACTOR) and use it when computing the new capacity.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,164 @@
+        if (index >= 0 && index <= size) {
+            for (int i = size; i > index; i--) {
+                elementData[i] = elementData[i - 1];
+            }
```
Lines 46-48: Insertion uses a manual loop to shift elements. The checklist requires: "Use [System.arraycopy()] to move your array elements." Replace this loop with System.arraycopy for correctness and clarity.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,164 @@
+                    + "positive length of an array");
+        }
+        elementData = EMPTY_ELEMENTDATA;
+    }
```
Lines 17-23: Constructor `ArrayList(int capacity)` currently ignores the provided capacity and sets `elementData = EMPTY_ELEMENTDATA`. This is incorrect — it should allocate `elementData = new Object[capacity]` when capacity > 0 (and handle zero/negative appropriately). This violates the requirement to behave like java.util.ArrayList (default capacity and respecting initial capacity). See task description: "Make your `ArrayList` behave the same way as `java.util.ArrayList` (it should have the default capacity, the internal array should grow 1.5 times when it is full, etc.)."

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,164 @@
     public void set(T value, int index) {
-
+        if (index >= 0 && index < size) {
+            elementData[index] = value;
```
Line 76-79: `public void set(T value, int index)` returns void. The task requires that set replace the element at index and return the previous element (ArrayList semantics). Update the method signature and implementation to return the previous element, and use centralized index validation. Also ensure the signature matches your project's List<T> interface.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,164 @@
+            for (int j = index; j < size(); j++) {
+                elementData[j] = elementData[j + 1];
+            }
+        } else {
```
Lines 88-96: `remove(int index)` decrements size and shifts elements but does not null out the vacated slot after shifting. After shifting, you must set the freed slot to null (e.g. `elementData[size] = null`) to avoid retaining references. Also prefer System.arraycopy for shifting rather than the loop. This matches expected ArrayList semantics and memory behavior.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,164 @@
+        // Сдвигаем элементы после найденного
+        for (int j = foundIndex; j < size - 1; j++) {
+            elementData[j] = elementData[j + 1];
+        }
```
Lines 93-95 and 135-137: Manual loops are used to shift elements during removals. Checklist requires: "Use [System.arraycopy()] to move your array elements." Replace these manual loops with System.arraycopy calls.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,164 @@
+
+        // Если элемент не найден — выбрасываем исключение
+        if (foundIndex == -1) {
+            throw new NoSuchElementException("Element not found");
```
Lines 52-53, 71-72, 81-82, 97-98, 131: Exception messages are vague (e.g. "Please enter the index of the element in the array" or "Element not found"). The checklist asks: "Make your exceptions informative." Update messages to include details like the invalid index and current size (e.g. "Index X out of bounds for size Y") and for missing element include the element or clearer context (e.g. "Element not found: " + element).

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,164 @@
 
     @Override
     public void add(T value, int index) {
 
```
Lines 36-38: Method `public void add(T value, int index)` — please confirm that this signature matches your project's List<T> interface. Standard Java List uses `add(int index, E element)`. If your List interface expects `add(int, T)` then this will not compile; either change the signature to match the interface or ensure the interface declares this order. Verify and align signatures with the List<T> interface used in the project.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,164 @@
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
```
Lines 58-62 (addAll): Current implementation repeatedly calls add(list.get(i)) which will repeatedly trigger growth checks. After you centralize grow/resize logic, consider sizing once up-front when possible (e.g. ensureCapacity(size + list.size())) to improve performance and mimic java.util.ArrayList behavior in spirit (optional optimization). This follows checklist guidance to avoid repeated code and expensive repeated resizes.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,164 @@
+    @Override
+    public String toString() {
+        Object[] actualElements = Arrays.copyOf(elementData, size);
+        return Arrays.toString(actualElements);
```
Lines 160-161: toString() uses Arrays.copyOf and Arrays.toString — again this uses java.util.Arrays which is disallowed by the checklist: "Don't use Objects, Arrays, or any other util class." Implement toString using a StringBuilder and iterate up to `size`, or create your own array copy using System.arraycopy if you need a temporary array.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

I’m requesting changes because several critical requirements are not met and they will block passing tests and violate the checklist.

- Critical blockers to fix:
  1) Import and usage of `java.util.Arrays` (Lines 3, 160-161, plus in resizing) — forbidden by the checklist. Replace `Arrays.copyOf` and `Arrays.toString` with your own logic using `System.arraycopy` and a manual `toString` via `StringBuilder`.
  2) Constructor `ArrayList(int capacity)` (Lines 17-23) ignores the provided capacity. It must allocate `elementData = new Object[capacity]` when capacity > 0; handle 0 and negative with correct behavior/messages.
  3) Duplicated resize logic and inline growth (Lines 27-31 and 39-43). Extract a single private `grow/resize` method that computes new capacity using a named constant like `GROWTH_FACTOR` and copies with `System.arraycopy`.
  4) Magic number `1.5` (Lines 29, 41). Introduce `private static final double GROWTH_FACTOR = 1.5` and use it.
  5) `set(T value, int index)` (Lines 76-79) must return the previous element. Update the signature and implementation to match your `List<T>` interface and ArrayList semantics.
  6) Shifting on insert/remove uses manual loops (Lines 46-48; 93-95; 135-137). Replace with `System.arraycopy` and after removals set the freed slot to `null` (e.g., `elementData[--size] = null`).
  7) Exception messages (Lines 52-53, 71-72, 81-82, 97-98, 131) should be informative, e.g., `Index X out of bounds for size Y`, `Illegal capacity: X`, or `Element not found: ...`.

Minor/optional improvements:
- Centralize index checks into helpers like `checkElementIndex(int)` and `checkPositionIndex(int)`.
- In `addAll`, pre-grow once (ensure capacity for `size + list.size()`).

You’ve covered much of the basic functionality—nice work. Addressing the above will align your implementation with java.util.ArrayList behavior and meet the checklist requirements.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

anurmetov committed:
Changes due to review:
- Changed whole toString() method on System.arraycopy princip
- Incapsulated add methods
- Added constants instead of hardcode numbers inside the code
- Edited a manual loop shifting on System.arraycopy shifting

anurmetov committed:
deleted old code

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,172 @@
+            for (int j = index; j < size(); j++) {
+                elementData[j] = elementData[j + 1];
+            }
+            elementData[size] = null;
```
Violation — manual element shift in remove(int):
This method uses a manual for-loop to shift elements after removal. The checklist explicitly requires: "Use [System.arraycopy()](https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#arraycopy-java.lang.Object-int-java.lang.Object-int-int-) to move your array elements." Replace the loop with a System.arraycopy call and null the freed slot after decrementing size to avoid memory leaks and match the checklist. Example pattern: `System.arraycopy(elementData, index + 1, elementData, index, size - index - 1); elementData[--size] = null;`

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,172 @@
+        elementData[size - 1] = null;
+
+        // Уменьшаем размер
+        size--;
```
Violation — manual element shift in remove(T element):
The `remove(T element)` method shifts elements using a loop. The checklist requires: "Use [System.arraycopy()](https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#arraycopy-java.lang.Object-int-java.lang.Object-int-int-) to move your array elements." Replace this loop with System.arraycopy and null out the freed slot after decrementing size. This will also reduce code duplication and improve performance. Also ensure the method decreases size only once (after shifting) and returns the removed element.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,172 @@
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds for size " + size);
+        }
```
Violation — repeated index validation logic (centralize checks):
You perform index checks inline in `add(T value, int index)`, `get(int)`, `set(T,int)` and `remove(int)`. The checklist instructs: "If the logic of your code repeats - move it to a separate private method. For example, the index checks for the `get()`, `set()`, and `remove()` methods should be moved to a separate method." Centralize index validation into helpers (e.g., `checkElementIndex(int)` for 0 <= index < size and `checkPositionIndex(int)` for 0 <= index <= size) and call them from all affected methods to ensure consistent behavior. This will remove duplicated code and make exceptions consistent.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,172 @@
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds for size " + size);
+        }
     }
```
Violation — `set` should return the previous element to match java.util.ArrayList semantics:
The project description requires that your ArrayList behave like `java.util.ArrayList` ("Make your `ArrayList` behave the same way as `java.util.ArrayList`"). In `java.util.ArrayList` the `set(int index, E element)` method returns the previous element at the index. Your `set(T value, int index)` currently returns void. Change the signature to `public T set(T value, int index)` and return the previous value after index validation. See task description: "Make your `ArrayList` behave the same way as `java.util.ArrayList` (it should have the default capacity, the internal array should grow 1.5 times when it is full, etc.)."

anurmetov replied:
@mateacademy-ai-mentor if i change the parameters order and return then test which were written by developer wont work and the List interface will not allow me it to do

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,172 @@
+        // Если элемент не найден — выбрасываем исключение
+        if (foundIndex == -1) {
+            throw new NoSuchElementException("Element not found" + element);
+        }
```
Violation — make exceptions informative:
The checklist requires: "Make your exceptions informative. If you throw an Exception, you should always add an informative message so that it would be clear from the stack trace what exactly went wrong." Currently `remove(T element)` throws `new NoSuchElementException("Element not found" + element);` — this concatenation lacks spacing and does not clearly represent null. Improve message formatting, e.g. `new NoSuchElementException("Element not found: " + String.valueOf(element))`. Apply the same clarity to other exception messages if needed.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

**Decision: REQUEST_CHANGES**

Your implementation is close, but there are several critical checklist and behavior violations that block approval.

Critical issues to fix:
- Use `System.arraycopy` for shifting on removals:
  - In `remove(int index)` (src/main/java/core/basesyntax/ArrayList.java), you currently shift elements with a loop. The checklist requires `System.arraycopy`. Replace the loop with: `System.arraycopy(elementData, index + 1, elementData, index, size - index - 1); elementData[--size] = null;` and return the removed element.
  - In `remove(T element)`, after finding the index to remove, perform the same `System.arraycopy`-based shift and then `elementData[--size] = null;`. Ensure size is decremented exactly once.
- Centralize index validation (duplicated in `add(index)`, `get`, `set`, `remove(int)`):
  - Extract helpers: `checkElementIndex(int index)` for `0 <= index < size` and `checkPositionIndex(int index)` for `0 <= index <= size` and use them consistently. This is an explicit checklist requirement and prevents inconsistent checks.
- Align `set` method with java.util.ArrayList semantics:
  - Change signature to `public T set(T value, int index)` and return the previous element at that index after validation.
- Make exception messages informative:
  - In `remove(T element)`, format as `new NoSuchElementException("Element not found: " + String.valueOf(element))` (note the colon and null-safe value). Review other thrown messages for clarity as well.

Minor improvements (nice to have, not blocking):
- Consider throwing `IllegalArgumentException` with message like `"Illegal capacity: " + capacity` for negative-capacity constructor, and using a shared empty array for zero capacity.
- Optionally pre-grow in `addAll` to avoid repeated resizes.

You’ve done well with constants, centralized resize logic, and using `System.arraycopy` in insertions. Address the above items to fully meet the checklist and java.util.ArrayList behavior. Keep going—you’re very close.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

anurmetov committed:
- changed manual shifting in remove methods to system.arraycope princip

anurmetov committed:
- changed validating implementation

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,170 @@
+    private int size = 0;
+
+    public ArrayList() {
+        elementData = EMPTY_ELEMENTDATA;
```
This violates the task description: "Make your ArrayList behave the same way as java.util.ArrayList (it should have the default capacity, the internal array should grow 1.5 times when it is full, etc.)." The no-arg constructor sets a zero-length array (`EMPTY_ELEMENTDATA`) instead of allocating the default capacity. Change the constructor to allocate `new Object[DEFAULT_CAPACITY]` so the internal storage begins with the default capacity. See the constructor implementation at elementData assignment.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,170 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
```
This violates the checklist item: "Don't begin class or method implementation with an empty line. Remove all redundant empty lines, be careful :)" There is an unnecessary empty line before the class declaration (blank line at line 4). Remove redundant blank lines at the top of the file to follow the checklist.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,170 @@
+
+    private boolean checkElementIndex(int index) {
+        return index >= 0 && index < size;
+    }
```
This violates the checklist item: "If the logic of your code repeats - move it to a separate private method. For example, the index checks for the `get()`, `set()`, and `remove()` methods should be moved to a separate method." You currently have `checkElementIndex(int)` returning boolean but still repeat throw logic in `get`, `set`, and `remove(int)` (remove(int) uses an inline `if (index >= 0 && index < size)`). Replace boolean helpers with private methods that perform validation and throw `ArrayListIndexOutOfBoundsException` with an informative message and call them from all index-based methods to remove duplication.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,170 @@
+            size--;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds for size " + size);
```
Index validation is inconsistent and duplicated in `remove(int)`. Instead of using the inline check `if (index >= 0 && index < size)` and duplicating the throw, call a centralized helper that throws `ArrayListIndexOutOfBoundsException` with the informative message `"Index " + index + " out of bounds for size " + size`. This matches the checklist guidance to avoid repeating index-checking logic. See the inline check at lines 95-102.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,170 @@
-
+        resizeCapacity();
+        // Сдвиг елементов вправо
+        if (!isValidAddIndex(index)) {
```
This violates the expected behaviour: `add(T value, int index)` calls `resizeCapacity()` before validating the index. Validate the index first (and throw the custom exception if invalid) and only then call `resizeCapacity()` and perform the array shift. This ensures capacity growth happens only for valid operations (growth only when needed). Move the index check so it runs before line 53 (`resizeCapacity()`).

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,170 @@
+            String message = (element == null)
+                    ? "Element 'null' not found"
+                    : "Element '" + element.toString() + "' not found";
+            throw new NoSuchElementException(message);
```
This relates to the checklist item: "Make your exceptions informative. If you throw an Exception, you should always add an informative message so that it would be clear from the stack trace what exactly went wrong." The `remove(T element)` method builds two different messages (special-case for null). Prefer a single, null-safe message pattern such as `new NoSuchElementException("Element not found: " + String.valueOf(element))` for consistency and clarity. Update the message creation at lines 131-135.

mateacademy-ai-mentor reviewed (approved):
### Review Summary

Educational approval: I’m approving this to help you progress, but there are a few important improvements to make next.

What you did well:
- Good separation of concerns with `resizeCapacity()` and use of `System.arraycopy`; correct 1.5× growth factor and proper nulling of vacated slots.

Critical fixes to address soon:
- Default capacity: the no-arg constructor initializes `elementData` to an empty array. Per the task, start with the default capacity (allocate `new Object[DEFAULT_CAPACITY]`).
- Centralize index checks: avoid repeated logic in `get`, `set`, and `remove(int)`. Create private void helpers that throw `ArrayListIndexOutOfBoundsException` with message: "Index " + index + " out of bounds for size " + size, and call them everywhere (lines ~95–102 show duplication in `remove(int)`).
- In `add(T value, int index)`, validate the index before calling `resizeCapacity()` so capacity only grows for valid operations (move the check above the call at around line 53).

Minor improvements:
- Use a single, null-safe `NoSuchElementException` message in `remove(T element)`, e.g., "Element not found: " + String.valueOf(element).
- Remove the redundant blank line at the top of the file and ensure constructors/methods don’t start with empty lines.

You’re very close—solid structure and growth logic are already in place. Tightening the constructors, index-checking helpers, and exception messages will align fully with the checklist and expected behavior. Keep it up!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,170 @@
+        } else {
+            elementData[index] = value;
+        }
     }
```
This violates the task requirement to behave like java.util.ArrayList. The class currently declares `public void set(T value, int index)` and does not return the previous element. The task description states: "Make your `ArrayList` behave the same way as `java.util.ArrayList` (it should have the default capacity, the internal array should grow 1.5 times when it is full, etc.)." You must change the signature to return the replaced element (e.g., `public T set(T value, int index)`), validate the index, store the old value, replace it, and return the old value.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,170 @@
+        }
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
```
resizeCapacity() is called before validating the insertion index. This can cause unnecessary growth when the provided index is invalid. Validate the index first (use `isValidAddIndex(index)`), and only call your grow/resize method after validation succeeds. Also note the checklist guidance to keep resizing logic separate: "Resize the array in a separate method."

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,170 @@
+            removedValue = (T) elementData[index];
+            System.arraycopy(elementData, index + 1, elementData, index, size - index - 1);
+            elementData[size - 1] = null;
+            size--;
```
The index check for `remove(int)` duplicates logic instead of using the helper method. The checklist explicitly recommends centralizing repeated logic: "If the logic of your code repeats - move it to a separate private method. For example, the index checks for the `get()`, `set()`, and `remove()` methods should be moved to a separate method." Use `checkElementIndex(index)` (or a helper that throws) to make checks consistent and avoid duplication.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,170 @@
+        elementData = EMPTY_ELEMENTDATA;
+    }
+
+    public ArrayList(int capacity) {
+        if (capacity < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + capacity
+                    + " out of bounds for size " + size);
+        } else {
+            elementData = new Object[capacity];
```
Constructor handles negative capacity by throwing `ArrayListIndexOutOfBoundsException` with an index-style message. The checklist requires informative exceptions: "Make your exceptions informative. If you throw an Exception, you should always add an informative message so that it would be clear from the stack trace what exactly went wrong." Consider throwing `IllegalArgumentException` with `"Illegal capacity: " + capacity` or at minimum change the message to indicate an invalid capacity argument.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,170 @@
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
```
`addAll` adds elements one-by-one, which may cause repeated resizes. For better performance, consider ensuring capacity for `size + list.size()` once before copying elements to avoid multiple reallocations (optional but recommended).

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,170 @@
     @Override
     public boolean isEmpty() {
-        return false;
+        return size <= 0;
```
`isEmpty()` returns `size <= 0`. Prefer `size == 0` to express intent clearly (size should not be negative). This is a readability suggestion.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,170 @@
+        for (Object element : actualElements) {
+            sb.append(element).append(" ");
+        }
+        return sb.toString();
```
toString() appends a trailing space after the last element. It's harmless but non-idiomatic. Consider formatting similar to java.util.ArrayList (brackets and commas) or at least avoid the trailing space.

mateacademy-ai-mentor reviewed (approved):
### Review Summary

Educational approval: moving you forward while highlighting key improvements to align with java.util.ArrayList behavior and the checklist.

What you did well:
- Clear constants for `DEFAULT_CAPACITY` and `GROWTH_FACTOR`, separate resize method, and solid use of `System.arraycopy` for shifts. 
- `remove(T)` correctly handles `null`, uses `equals`, and throws `NoSuchElementException` when not found.

Critical fixes to address next:
- At src/main/java/core/basesyntax/ArrayList.java: change `public void set(T value, int index)` to `public T set(T value, int index)`, validate the index, store the old value, replace it, and return the old value (semantic parity with `java.util.ArrayList`).
- In `add(T value, int index)`, validate the index first (e.g., `checkPositionIndex(index)`), then grow if needed; currently resize happens before validation which can grow on invalid indices.
- DRY index checks: refactor `remove(int)` to use a centralized throwing helper (e.g., `checkElementIndex(index)`) instead of inline conditions; apply the same helper for `get/set/remove` per checklist.

Recommended improvements:
- Constructor with negative capacity: throw `IllegalArgumentException("Illegal capacity: " + capacity)` for a clearer message/type.
- Optionally pre-ensure capacity in `addAll` to avoid multiple resizes; prefer `size == 0` in `isEmpty()`; tidy `toString()` to avoid trailing space.

Nice progress—these targeted tweaks will make your implementation both correct and clean.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

anurmetov committed:
- changed code due to review

anurmetov committed:
- changed code due to review

Elena-Bruyako commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,162 @@
+
+    }
+
+    private void resizeCapacity() {
```
private methods should be placed after all public methods
check other places

Elena-Bruyako commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,162 @@
+        }
+    }
+
+    // get, set, remove
```
remove comment

Elena-Bruyako commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,162 @@
+        }
+    }
+
+    // add (index, value)
```
remove comment

Elena-Bruyako commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,162 @@
-
+        checkElementAddIndex(index);
+        resizeCapacity();
+        // Сдвиг елементов вправо
```
remove comment

Elena-Bruyako commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,162 @@
     public T remove(int index) {
-        return null;
+        checkElementAccessIndex(index);
+
```
remove empty line
```suggestion
```

Elena-Bruyako commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,162 @@
     @Override
     public T remove(T element) {
-        return null;
+        T removedValue = null;
```
```suggestion
        int index = index(element);
        if (index == -1) {
            throw new NoSuchElementException("There is no such element in the list, " + element);
        }
        return remove(index);
```

Elena-Bruyako reviewed (changes requested):
Good job! Let's improve your solution
